src/blockchain.rs

```diff
@@ -217,86 +219,131 @@ impl Blockchain {
+            println!("INCORRECT TIMESTAMPS");
+            return false;
+        }
+
+        let golden_ticket_count = block.transactions().iter().fold(0, |acc, tx| {
+            if tx.core.broadcast_type() == &TransactionType::GoldenTicket {
+                acc + 1
+            } else {
+                acc
+            }
+        });
+        if golden_ticket_count > 1 {
+            println!("TOO MANY GOLDEN TICKETS");
             return false;
         }
 
         let transactions_valid = block
             .transactions()
             .iter()
-            .all(|tx| self.validate_transaction(block, tx, fork_tuple));
+            .all(|tx| self.validate_transaction(previous_block, block, tx, fork_tuple));
 
         transactions_valid
     }
 
     fn validate_transaction(
         &self,
+        previous_block: &Block,
         block: &Block,
         tx: &Transaction,
         fork_tuple: &ForkTuple,
     ) -> bool {
-        // check that our sigs are valid
-        if let Some(output_slip) = self.utxoset.output_slip_from_slip_id(&tx.core.inputs()[0]) {
-            // let serialized_tx: Vec<u8> = tx.core.clone().into();
-            // make_message_from_bytes(&serialized_tx[..]);
-            let hash_message = tx.core.hash();
-            if !verify_message_signature(&hash_message, &tx.signature(), &(output_slip.address())) {
-                return false;
-            };
-
-            // validate our slips
-            let inputs_are_valid = tx.core.inputs().iter().all(|input| {
-                return !self
-                    .utxoset
-                    .is_slip_spent_at_block(input, block, fork_tuple);
-            });
-
-            if !inputs_are_valid {
-                return false;
-            }
+        match tx.core.broadcast_type() {
+            TransactionType::Normal => {
+                if tx.core.inputs().len() == 0 && tx.core.outputs().len() == 0 {
+                    return true;
+                }
 
-            // valuidate that inputs are unspent
-            let input_amt: u64 = tx
-                .core
-                .inputs()
-                .iter()
-                .map(|input| {
-                    self.utxoset
-                        .output_slip_from_slip_id(input)
-                        .unwrap()
-                        .amount()
-                })
-                .sum();
-
-            let output_amt: u64 = tx.core.outputs().iter().map(|output| output.amount()).sum();
-
-            if input_amt != output_amt {
-                return false;
+                if let Some(output_slip) =
+                    self.utxoset.output_slip_from_slip_id(&tx.core.inputs()[0])
+                {
+                    let hash_message = tx.core.hash();
+                    if !verify_message_signature(
+                        &hash_message,
+                        &tx.signature(),
+                        &(output_slip.address()),
+                    ) {
```

Comment:
It's not immediately obvious that we need to get an input in order to validate the signature. I think some comment here could be very helpful. There is a stub function in utxo_set called get_receiver_for_slips. Please use that instead of output_slip_from_slip_id. The idea is that all the inputs need to have the same receiver and utxoset has this information. if it returns None, then the tx is invalid. This also avoids the confusion of "output_slip" caused because output_slip is a bit confusing to read here because it's an output from another transaction.